Added Bone Extraction Mode: Expression

emil-dietrich opened the pull request:
Each bone is extracted if the expression returns `True`.
Each bone is accessible in the expression by the var `bone`.

- [x] TODO: compile the expression once and use the bytecode with `eval(bytecode, ...)`

emil-dietrich committed:
Added Bone Extraction Mode: Expression
Each bone is extracted if the expression returns `True`.
Each bone is accessible in the expression by the var `bone`.

TODO: compile the expression once and use the bytecode with
`eval(bytecode, ...)`

BlenderBoi commented:
Interesting contribution! I will take some time to review the code and test the feature, hopefully it will be a good addition to Game Rig Tools

emil-dietrich committed:
Implement extract mode to filter expression
extract mode cases are handled once, not once per iteration
custom expression is compiled once, not once per iteration

emil-dietrich committed:
Better error origin in case of invalid extract mode expression

emil-dietrich commented:
For the last commit [a8414a8](https://github.com/BlenderBoi/Game_Rig_Tools/pull/3/commits/a8414a8a5724e5787dafe1aea1b5f816b7a2f1cb) 
before:
![image](https://github.com/BlenderBoi/Game_Rig_Tools/assets/148558055/7296258a-a52a-436b-9413-b1afd9abdf69)
after:
![image](https://github.com/BlenderBoi/Game_Rig_Tools/assets/148558055/c9b23586-37f6-45d2-adfd-2d2ee18a8e87)

`xx` was what I entered in the input field

BlenderBoi commented:
I will need some time to check if this is safe

eval seems to have some security risk, I will need some time to research if this is ok to add

emil-dietrich commented:
If there is a security risk its that you can use python within blender.
Everyone can open a text file in blender and press execute.
To add to it: Only one variable is accessible by the expression and that is `bone`

BlenderBoi commented:
Fair Point, I will do some testing and if it works will I will merge it

emil-dietrich commented:
Dont merge what you dont understand. Dont feel pressured to add this before you are comfortable with the changes.

emil-dietrich commented:
Sry, for another addition:
rigify also dynamically executes python code, see `finalize_script` property.
It calls it as follows (in generate.py:594): `exec(finalize_script.as_string(), {})`

_(You can see where rigify is located on your hard drive if you go to the preferences -> addons -> rigify -> File: ...)_

`exec` and `eval` are basically the same, but `eval` returns a value and `exec` does not.

The above still applies - if you dont feel comfortable with the change im fine with that.

BlenderBoi commented:
I think it would be a good addition, I will merge it, but I will take some time to study it
